Add sensei home rest API mock.

aaronfc opened the pull request:
Fixes #5676 

### Changes proposed in this Pull Request
* Create `/sensei-internal/v1/home` endpoint to return all the data needed by new Sensei Home screen.
* This is only an initial pull request to discuss the interface to be consumed from Sensei Home frontend. Implementation will come later probably in separate tickets.
* Based on 1-gh-Automattic/senseilms-com-plugins

### Testing instructions
- Check the structures fit all requirements in the Home i2 iteration P2.
- Check pull request for comments with further questions/doubts.

### Discussion
- ~Should we return from the backend the entries for the Help section? They could be frontend only since we won't have a place to persist those entries and will be basically hardcoded in the backend.~ Yes, returning from the backend so we can customise them based on plugin installed and extensions.
- We might end up separating the Tasks List into a separate endpoint or maybe even other parts too. Any ideas on granularity here?
- ~Task list also contains an image that changes based on the current step. Should we return the image url from the backend or maybe let the frontend decide the image by checking which is the first "not done" task from the list? The second part would mean returning some ID so frontend can identify the step and map to a correct image. Probably it's easier to return the image url from backend. WDYT?~ For the moment returning an optional image on every task.

aaronfc committed:
Add mocked rest API to be consumed from Home.

- New endpoint returning all the information needed by the Home API.
- Current implementation is a mock so we can start discussing on the
  actual API to be used from frontend.

aaronfc commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,216 @@
+					'image'        => 'http://senseilms.com/wp-content/uploads/2022/02/sensei-post-to-course-80x80.png',
+					'description'  => 'Turn your blog posts into online courses.',
+					'price'        => 0,
+					'is_featured'  => false, // To be used to differentiate between featured and not featured extensions.
```
Featured and non-featured extensions are all returned in a single collection but can be differentiated by `is_featured` property.

fjorgemota replied:
Unless I'm missing something obvious, are we going to have featured extensions?

I remember that we had those on the extensions, but I'm not 100% sure about the new Sensei Home. I think Sensei Pro is the only extension that we really want to feature more... :thinking:

fjorgemota replied:
Hey, just as FYI, I just got a reply from Ronnie confirming that we won't need featured extensions support: p6rkRX-4so-p2#comment-4848

aaronfc replied:
Ok, thanks for looking into that! Removed here 7a7dfd105

And added support for the promotional banner in a very simple way here 1a725d3dc
(I have added more details/doubts about the promotional banner here)

aaronfc commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,216 @@
+					'url'   => '???',
+				],
+			],
+			'quick_links' => [
+				// TODO: Generate or hardcode in frontend.
+				'Courses'           => [
```
Not sure about this structure. Is made to fit the usage from frontend, but I was tempted to separate categories from links:
```
'quick_links' => [
  'categories' => ['id' => 'courses', title => 'Courses'],
  'links' => [ ['title' => '...', 'category' => 'course', 'url' => '...'], ....],
  ]
```

Any strong opinions?

jom replied:
In the implementation the section titles will need to be i11n-able. Although I doubt it would cause issues, part of me thinks that the title shouldn't come from the key. I'd say either your solution or something like:
```php
'quick_links' => [
	[
		'title' => 'Courses',
		'items' => [
			[
				'title' => 'Import a Course',
				'url'   => 'http://...',
			],
			// ...
		],
	],
	// ...
]
```

fjorgemota replied:
I also think that having the title as a property instead as being the key is better.

I like Jake's solution here a little more, personally, as it avoids the need to have an extra lockup/filter to get the links from a given category.

aaronfc replied:
Changed to what @jom suggested in 71b27dbe3

aaronfc commented on tests/unit-tests/rest-api/test-class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,72 @@
+ *
+ * @group rest-api
+ */
+class Sensei_REST_API_Home_Controller_Tests extends WP_Test_REST_TestCase {
```
Very basic test for permissions, will extend as the implementation grows.

aaronfc committed:
Extend response for tasks to include image url.

jom commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,219 @@
+				'items'    => [
+					[
+						'title' => 'Introducing Interactive Videos For WordPress',
+						'date'  => '2022-08-31T21:41:38',
```
Just a wee note that this is the format (RFC3339) used on the SenseiLMS.com endpoint but we'll want to localize it (probably just the date and ignore the time) before sending to frontend.

aaronfc replied:
Clarified in d5d096198

jom reviewed (commented):
Hi! Some initial thoughts. Did we decide to put notices in this one or another endpoint?

> Should we return from the backend the entries for the Help section? They could be frontend only since we won't have a place to persist those entries and will be basically hardcoded in the backend.

Part of me thinks we should pass them in this endpoint; that way we can check for the license and show the appropriate `Create a support ticket` link.

> Task list also contains an image that changes based on the current step.

I don't have a strong preference. Seems easy enough for the backend to figure it out and it shouldn't change in the frontend after loading. I am also fine if we decide to just send the image for each step and let the frontend decide which one to show.

fjorgemota commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,219 @@
+				],
+			],
+			'help'        => [
+				// TODO: Generate or hardcode in frontend.
```
I'm not sure about hardcoding this on the frontend, as we'll need to verify whether the license is valid first. May as well leave it dynamic and coming from the backend, as Jake pointed out.

aaronfc replied:
Agree! Added skeleton in 145cee198.

fjorgemota reviewed (commented):
This is looking good. I left some extra cmments.

By the way, you're probably already on it, but as suggested on the team meeting I'd suggest splitting the different data returned by this endpoint a little more.

It might as well be a good idea to add a filter on the output returned by the `get_data`, and maybe we could even add new callbacks to that filter to just add the specific data? (e.g, run `add_filter('sensei_home_api_data', [$this, 'add_tasks_list_data']);`, etc.)

I'm not sure, I'm just giving out some ideas, which might or might not be good. The main idea around the filter is to allow Sensei Pro to hook on it and add licensing information, but this is just another idea that I had while thinking about the API, hehe.

aaronfc committed:
Move tasks image as optional inside each task item.

aaronfc committed:
Update with `feature/` branch.

aaronfc committed:
Improve Quick Links categorization structure.

aaronfc committed:
Add mockup for help section.

aaronfc committed:
Localize dates for news.

aaronfc committed:
Remove `is_featured` mark for extensions.

aaronfc committed:
Add `show_sensei_pro_promo` boolean parameter.

aaronfc commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,255 @@
+					'more_url'     => 'http://senseilms.com/product/sensei-lms-post-to-course-creator/',
+				],
+			],
+			'show_sensei_pro_promo' => false, // Whether we should show the promotional banner for Sensei Pro or not..
```
👀  @Automattic/evergreen

I thought about returning text/images from the backend for the promotion but it looks too cumbersome. I considered also returning som kind of `active_promotion` attribute that we can populate with different keys (that frontend must know about).
But finally decided to go the simplest way: return a boolean specific for this case and we can iterate this API to do something more generic if we have other use cases in the future.

WDYT?

jom replied:
I think just having a boolean is fine. Everything else can be kept pretty statically in the component we use for the Sensei Pro promo.

fjorgemota replied:
While it could be cool to have different "ads", I feel like this isn't needed at the first moment. Maybe it's worth asking Ronnie/Andrei and implementing it as a low-priority task if needed, though.

aaronfc committed:
Add notifications to Home Screen REST API.

aaronfc commented:
Added support for notifications in here 26a23f452, please @jom take a look.
Not sure how the dismiss action should behave, maybe we need an endpoint to mark a notification as dismissed and include some kind of notification id?

aaronfc committed:
Add list of pending activation products.

aaronfc commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,284 @@
+				],
+			],
+			'show_sensei_pro_promo'       => false, // Whether we should show the promotional banner for Sensei Pro or not.
+			'notifications'               => [
+				[
+					'title'            => 'Your Sensei Pro license expires on 12.09.2022.',
+					'main_action'      => [
+						'title' => 'Update now',
+						'url'   => 'https://...',
+					],
+					'secondary_action' => [
+						'title' => 'What\'s new',
+						'url'   => 'https://...',
+					],
+					'level'            => 'error', // One of: info, warning, error.
+					'dismissible'      => false, // The default value is true.
+				],
+				[
+					'title'            => 'Good news, reminder to update to latest version',
+					'main_action'      => [
+						'title' => 'Update now',
+						'url'   => 'https://...',
+					],
+					'secondary_action' => [
+						'title' => 'Link for more information',
+						'url'   => 'https://...',
+					],
+					'level'            => 'info', // One of: info, warning, error.
+					'dismissible'      => true, // The default value is true.
+				],
+			],
```
This is new in order to model notifications.

aaronfc commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,284 @@
+					'dismissible'      => true, // The default value is true.
+				],
+			],
+			'products_pending_activation' => [ 'sensei-pro' ],
```
This part has been added to model the form that will allow for the activation of Sensei Pro license.

fjorgemota replied:
While I know that this is only a mockup, I'm not sure about putting this data here, to be honest.

With that said, I might be wrong, but.. .do we need to know which products are pending activation?  I feel like a boolean or even an object containing keys like `activated`, `product_name` (if needed, but not sure), and `expiration_date` is already enough. 

I wonder if there's a situation where we might have more than one product pending activation..

aaronfc replied:
> While I know that this is only a mockup, I'm not sure about putting this data here, to be honest.

I don't understand this. Do you mean that you would rather have a separate endpoint?

> With that said, I might be wrong, but.. .do we need to know which products are pending activation? I feel like a boolean or even an object containing keys like activated, product_name (if needed, but not sure), and expiration_date is already enough.

As far as my understanding a single boolean wouldn't work because we might want to differentiate between sensei-pro and interactive-blocks at least. Maybe we can keep it like this and once we start implementing the actual licensing part decide better on this (to be honest haven't looked too much into it).

fjorgemota replied:
> I don't understand this. Do you mean that you would rather have a separate endpoint?

Nah, it's only that I think this should be present on Sensei Pro instead, being added by a filter or something.

> As far as my understanding a single boolean wouldn't work because we might want to differentiate between sensei-pro and interactive-blocks at least. Maybe we can keep it like this and once we start implementing the actual licensing part decide better on this (to be honest haven't looked too much into it).

Fair enough.

fjorgemota commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,284 @@
+				],
+				[
+					'title'            => 'Good news, reminder to update to latest version',
+					'main_action'      => [
+						'title' => 'Update now',
+						'url'   => 'https://...',
+					],
+					'secondary_action' => [
+						'title' => 'Link for more information',
+						'url'   => 'https://...',
+					],
```
WDYT about having only one array of `actions` per notice, containing `title`, `url`, and maybe a `variant` (to define the style of the action, whether it's a link, a primary, or a secondary action)?

aaronfc replied:
I thought about that, but given that the variant will mark also the position I think it would be easier from the frontend to model them separately. I doubt we would get more than main and an optional secondary one.

fjorgemota commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,284 @@
+					'title'        => 'Sensei LMS Post to Course Creator',
+					'image'        => 'http://senseilms.com/wp-content/uploads/2022/02/sensei-post-to-course-80x80.png',
+					'description'  => 'Turn your blog posts into online courses.',
+					'price'        => 0,
```
I think this can be a string instead to avoid eventual problems with floats, so the prefix is already included. WDYT?

The extensions API already does that, for reference:

![image](https://user-images.githubusercontent.com/529864/191876055-ef9acb87-374e-4c0c-9df0-19aab5e7b7d3.png)

aaronfc replied:
Well, I was assuming here cents. And localising it would make it difficult for the frontend to identify the "free" scenario which requires a special install button text.

fjorgemota replied:
hmmm, good point. And not sure how to handle it, to be honest.

fjorgemota commented on includes/rest-api/class-sensei-rest-api-home-controller.php:
```diff
@@ -0,0 +1,284 @@
+	/**
+	 * Register GET / endpoint.
+	 */
+	public function register_get_data_route() {
```
Given that this is only one endpoint, do we need this additional method?

aaronfc replied:
I will remove if finally we have only a single endpoint :)

fjorgemota reviewed (commented):
>  Not sure how the dismiss action should behave, maybe we need an endpoint to mark a notification as dismissed and include some kind of notification id?

Having some sort of notification ID looks necessary, yeah. About the extra endpoint, it will depend on whether we want the notices to be hidden globally or locally. If locally, even `localStorage` might be already good enough.

----

With that said, this is looking good. I left some additional comments on the PR. :)

aaronfc committed:
Reformulating actions and link for notifications.

aaronfc commented:
After talking with @jom I have updated a little bit the notifications part. See here 5dd05951e

aaronfc committed:
Rename notification attributes.

- Renaming notification attributes to better match the schema from
  SenseiLMS.com notifications.
- Renamed also button/actions to use label instead of title.

jom reviewed (approved):
Looks good! I think any further changes we can make as we go.

fjorgemota reviewed (commented):
I know this is already merged, just wanted to reply two of your comments. :)
